common/src/ipc/serde.rs

```diff
@@ -0,0 +1,631 @@
+        Self { buf, pos: 0 }
+    }
+
+    /// Extract buffer back, discarding all internal state.
+    pub fn finish(self) -> T {
+        self.buf
+    }
+}
+
+impl<'a> Cursor<&'a mut [u8]> {
+    fn write(&mut self, bytes: &[u8]) {
+        let next = self.pos + bytes.len();
+        self.buf[self.pos..next].copy_from_slice(bytes);
+        self.pos = next;
+    }
+
+    fn write_tagged(&mut self, bytes: &[u8]) {
+        bytes.tag().serialize(self);
+        self.write(bytes);
+    }
+}
+
+impl<'a> Cursor<&'a [u8]> {
+    fn read(&mut self, count: usize) -> &'a [u8] {
+        let read = &self.buf[self.pos..self.pos + count];
+        self.pos += count;
+        read
+    }
+
+    fn read_tagged(&mut self) -> &'a [u8] {
+        let count = u32::deserialize(self);
+        self.read(count as usize)
+    }
+}
+
+trait Tagged<'a> {
+    fn tag(&self) -> u32;
+}
+
+impl<T> Tagged<'_> for [T] {
+    fn tag(&self) -> u32 {
+        self.len() as u32
+    }
+}
+
+impl Tagged<'_> for str {
+    fn tag(&self) -> u32 {
+        self.as_bytes().tag()
+    }
+}
+
+/// Serializes data structure into byte slice.
+pub trait Serialize {
+    /// Write self into buffer.
+    ///
+    /// # Panics
+    ///
+    /// If [`size`](Serialize::size) is incorrectly implemented, buffer can be
+    /// too small to fit all bytes.
+    fn serialize(&self, buffer: &mut Cursor<&mut [u8]>);
+
+    /// Calculate resulting size of serialized data structure.
+    ///
+    /// Resulting value will be used to allocate buffer, so incorrect implementation
+    /// will either result in panic after out of bounds access or wasted space.
+    fn size(&self) -> usize;
+
+    /// Same as [`size`](Serialize::size), but with space for slice length (now [`u32`])
+    #[doc(hidden)]
+    fn size_tagged(&self) -> usize {
+        size_of::<u32>() + self.size()
+    }
+}
+
+/// Deserializes byte slice into desired data structure.
+pub trait Deserialize<'a> {
+    /// Read bytes from buffer and interpret as desired.
+    ///
+    /// # Panics
+    ///
+    /// If provided buffer is messed up (i.e. it isn't filled by corresponding
+    /// [`Serialize`] implementation).
+    ///
+    /// # Safety
+    ///
+    /// This function isn't marked as unsafe, so in no circumstances can
+    /// implementations count on provided buffer to be well formed. Panic instead.
+    fn deserialize(buffer: &mut Cursor<&'a [u8]>) -> Self;
+}
+
+macro_rules! primitive {
+    ($($type:ty),+ $(,)?) => {
+        $(
+        impl Serialize for $type {
+            fn serialize(&self, buffer: &mut Cursor<&mut [u8]>) {
+                self.to_ne_bytes().serialize(buffer)
+            }
+
+            fn size(&self) -> usize {
+                size_of::<$type>()
+            }
+        }
+
+        impl Deserialize<'_> for $type {
+            fn deserialize(buffer: &mut Cursor<&[u8]>) -> Self {
+                let array = buffer.read(size_of::<$type>()).try_into().expect("slice is correctly sized");
```

Comment:
I believe this panic message is wrong. If the slice is incorrectly sized, it will panic in `read`, not here. This will fail if the conversion from `try_into` fails.

Reply:
`TryInto` will fail if `slice.len()` isn't equal to number of elements in requested array. This is why I expect slice to have correct size.

Reply:
You are right, my mistake.